src/decode/ykpt/packet_parser/packets.rs

```diff
@@ -0,0 +1,355 @@
+//! Intel PT packets and their constituents.
+
+use deku::prelude::*;
+use std::convert::TryFrom;
+
+/// The `IPBytes` field common to all IP packets.
+///
+/// This tells us what kind of compression was used for a `TargetIP`.
+#[derive(Clone, Copy, Debug, DekuRead)]
+pub(in crate::decode::ykpt) struct IPBytes {
+    #[deku(bits = "3")]
+    val: u8,
+}
+
+impl IPBytes {
+    #[cfg(test)]
+    pub(in crate::decode::ykpt) fn new(val: u8) -> Self {
+        debug_assert!(val >> 3 == 0);
+        Self { val }
+    }
+    /// Returns `true` if we need the previous TIP value to make sense of the new one.
+    pub(super) fn needs_prev_tip(&self) -> bool {
+        match self.val {
+            0b001 | 0b010 | 0b100 => true,
+            _ => false,
+        }
+    }
+}
+
+/// The `TargetIP` fields in packets which update the TIP.
+///
+/// This is a variable-width field depending upon the value if `IPBytes` in the containing packet.
+#[derive(Debug, DekuRead)]
+#[deku(id = "ip_bytes_val", ctx = "ip_bytes_val: u8")]
+pub(in crate::decode::ykpt) enum TargetIP {
+    #[deku(id = "0b000")]
+    OutOfContext,
+    #[deku(id = "0b001")]
+    Ip16(u16),
+    #[deku(id = "0b010")]
+    Ip32(u32),
+    #[deku(id_pat = "0b011 | 0b100")]
+    Ip48(#[deku(bits = "48")] u64),
+    #[deku(id = "0b110")]
+    Ip64(u64),
+}
+
+impl TargetIP {
+    #[cfg(test)]
+    pub(in crate::decode::ykpt) fn from_bits(bits: u8, val: u64) -> Self {
+        match bits {
+            0 => Self::OutOfContext,
+            16 => Self::Ip16(u16::try_from(val).unwrap()),
+            32 => Self::Ip32(u32::try_from(val).unwrap()),
+            48 => Self::Ip48(val),
+            64 => Self::Ip64(val),
+            _ => panic!(),
+        }
+    }
+
+    /// Decompress a `TargetIP` and `IPBytes` pair into an instruction pointer address.
+    ///
+    /// Returns `None` if the target IP was "out of context".
+    pub(in crate::decode::ykpt) fn decompress(
+        &self,
+        ip_bytes: IPBytes,
+        prev_tip: Option<usize>,
+    ) -> Option<usize> {
+        let res = match ip_bytes.val {
+            0b000 => {
+                debug_assert!(matches!(self, Self::OutOfContext));
+                return None;
+            }
+            0b001 => {
+                // The result is bytes 63..=16 from `prev_tip` and bytes 15..=0 from `ip`.
+                if let Self::Ip16(v) = self {
+                    prev_tip.unwrap() & 0xffffffffffff0000 | usize::try_from(*v).unwrap()
```

Comment:
Is this meant to be `usize` or `u64`?

Reply:
I knew you'd flag this :)

I chose `u64` for the compressed address and `usize` for the decompressed address.

Justification:
 - Before decompression the address is just a field in a packet. It wouldn't be right to call that an address, so usize makes little sense there.
 - After decompression we have a `RIP` (instruction pointer register) value, and it's also a virtual address. This is by definition a pointer-sized integer and `usize` is perfect. In theory, PT works in 32-bit mode (on a 64-bit machine) too, so there you wouldn't want a `u64`.

Do you buy it?

Reply:
I do!